packages/angular/cli/src/command-builder/schematics-command-module.ts

```diff
@@ -238,36 +243,49 @@ export abstract class SchematicsCommandModule
     return (this._workflowForExecution = workflow);
   }
 
-  private _defaultSchematicCollection: string | undefined;
-  protected async getDefaultSchematicCollection(): Promise<string> {
-    if (this._defaultSchematicCollection) {
-      return this._defaultSchematicCollection;
+  private _schematicsCollections: Set<string> | undefined;
+  protected async getschematicsCollections(): Promise<Set<string>> {
+    if (this._schematicsCollections) {
+      return this._schematicsCollections;
     }
 
-    let workspace = await getWorkspace('local');
+    const getschematicsCollections = (
+      configSection: Record<string, unknown> | undefined,
+    ): Set<string> | undefined => {
+      if (!configSection) {
+        return undefined;
+      }
 
-    if (workspace) {
-      const project = getProjectByCwd(workspace);
-      if (project) {
-        const value = workspace.getProjectCli(project)['defaultCollection'];
-        if (typeof value == 'string') {
-          return (this._defaultSchematicCollection = value);
-        }
+      const { schematicsCollections, defaultCollection } = configSection;
+      if (Array.isArray(schematicsCollections)) {
+        return new Set(schematicsCollections);
+      } else if (typeof defaultCollection === 'string') {
+        return new Set([defaultCollection]);
       }
 
-      const value = workspace.getCli()['defaultCollection'];
-      if (typeof value === 'string') {
-        return (this._defaultSchematicCollection = value);
+      return undefined;
+    };
+
+    const localWorkspace = await getWorkspace('local');
+    if (localWorkspace) {
+      const project = getProjectByCwd(localWorkspace);
+      if (project) {
+        const value = getschematicsCollections(localWorkspace.getProjectCli(project));
+        if (value) {
+          return (this._schematicsCollections = value);
+        }
       }
     }
 
-    workspace = await getWorkspace('global');
-    const value = workspace?.getCli()['defaultCollection'];
-    if (typeof value === 'string') {
-      return (this._defaultSchematicCollection = value);
+    const globalWorkspace = await getWorkspace('global');
+    const value =
+      getschematicsCollections(localWorkspace?.getCli()) ??
+      getschematicsCollections(globalWorkspace?.getCli());
+    if (value) {
+      return (this._schematicsCollections = value);
     }
 
-    return (this._defaultSchematicCollection = DEFAULT_SCHEMATICS_COLLECTION);
+    return (this._schematicsCollections = new Set([DEFAULT_SCHEMATICS_COLLECTION]));
```

Comment:
Nit: I see that this was pre-existing, but I think we should separate out these assignments into separate statements to avoid a side-effectful return.